### PR TITLE
Load file config YAML using core schema, ensure that env var substiut…

### PR DESCRIPTION
**Pull request by jack-berg**
…tion retains string types

Reflects a couple of recent spec PRs:
- https://github.com/open-telemetry/opentelemetry-specification/pull/3973
- https://github.com/open-telemetry/opentelemetry-specification/pull/3975

### CODE REVIEW

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java [hunk 1]
```diff
@@ -146,51 +152,93 @@ static Object loadYaml(InputStream inputStream, Map<String, String> environmentV
   private static final class EnvSubstitutionConstructor extends StandardConstructor {
 
     // Yaml is not thread safe but this instance is always used on the same thread
```
Comment (laurit) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java:154]: is this comment still valid?

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java [hunk 2]
```diff
@@ -146,51 +152,93 @@ static Object loadYaml(InputStream inputStream, Map<String, String> environmentV
...
-        }
-        entry.setValue(yaml.load(newVal.toString()));
+      return mapping;
+    }
+
+    private Object constructValueObject(Node node) {
+      if (!(node instanceof ScalarNode)) {
+        return super.constructObject(node);
+      }
+      Object value = super.constructObject(node);
```
Comment (laurit) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java:212]: I'd move this to the start of the method and use `value` in the preceding return statement.

**Review by laurit — APPROVED** [review]

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java [hunk 3]
```diff
@@ -146,51 +152,93 @@ static Object loadYaml(InputStream inputStream, Map<String, String> environmentV
...
-      // 3. Re-parse substituted value so it has correct type (i.e. yaml.load(newVal))
-      for (Map.Entry<Object, Object> entry : result.entrySet()) {
-        Object value = entry.getValue();
-        if (!(value instanceof String)) {
-          continue;
+      Map<Object, Object> mapping = settings.getDefaultMap().apply(node.getValue().size());
+      List<NodeTuple> nodeValue = node.getValue();
+      for (NodeTuple tuple : nodeValue) {
+        Node keyNode = tuple.getKeyNode();
+        Node valueNode = tuple.getValueNode();
```
Comment (breedx-splk) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java:178]: Maybe move the `valueNode` down closer to where it's used.

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java [hunk 4]
```diff
@@ -146,51 +152,93 @@ static Object loadYaml(InputStream inputStream, Map<String, String> environmentV
...
-        if (offset != val.length()) {
-          newVal.append(val, offset, val.length());
-        }
-        entry.setValue(yaml.load(newVal.toString()));
+      return mapping;
+    }
+
+    private Object constructValueObject(Node node) {
+      if (!(node instanceof ScalarNode)) {
+        return super.constructObject(node);
```
Comment (breedx-splk) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java:210]: Don't really need that `super` here.

```suggestion
        return constructObject(node);
```

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java (hunk 2 above)
Comment (breedx-splk) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java:212]:
```suggestion
      Object value = constructObject(node);
```

Comment (breedx-splk) [reply sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java:212]: Yeah it reads weirdly to have the thing constructed in both cases (if and implied else).

File: sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigurationParseTest.java [hunk 5]
```diff
@@ -432,7 +453,13 @@ private static java.util.stream.Stream<Arguments> envVarSubstitutionArgs() {
...
             mapOf(entry("key1", mapOf(entry("${STR_1}", "value1"))))),
         Arguments.of(
-            "key1:\n - ${STR_1}\n", mapOf(entry("key1", Collections.singletonList("${STR_1}")))));
+            "key1:\n - ${STR_1}\n", mapOf(entry("key1", Collections.singletonList("${STR_1}")))),
+        // Quoted environment variables
+        Arguments.of("key1: \"${HEX}\"\n", mapOf(entry("key1", "0xdeadbeef"))),
+        Arguments.of("key1: \"${STR_1}\"\n", mapOf(entry("key1", "value1"))),
+        Arguments.of("key1: \"${BOOL}\"\n", mapOf(entry("key1", "true"))),
+        Arguments.of("key1: \"${INT}\"\n", mapOf(entry("key1", "1"))),
+        Arguments.of("key1: \"${FLOAT}\"\n", mapOf(entry("key1", "1.1"))));
```
Comment (breedx-splk) [inline sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigurationParseTest.java:462]: It's edge-casey for sure, but for completeness I'd like to see a test where the value is replaced with the empty string. I could see a user down the road wanting to turn something off or just quiet some replacement string with an empty.

**Review by breedx-splk — APPROVED** [review]
Yeah this is great. I had a few nitpick/small comments, but good work overall. 🥂

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java (hunk 4 above)
Comment (jack-berg) [reply sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfiguration.java:210]: Good catch. At one point the implementation overrode `constructObject` so super was necessary but I adjusted it in a revision.

File: sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigurationParseTest.java (hunk 5 above)
Comment (jack-berg) [reply sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigurationParseTest.java:465]: Good call
